README.md
# `uniprot.rs` [![Star me](https://img.shields.io/github/stars/althonos/uniprot.rs.svg?style=social&label=Star&maxAge=3600)](https://github.com/althonos/uniprot.rs/stargazers)

*Rust data structures and parser for the [UniprotKB database(s)].*

[UniprotKB database(s)]: https://www.uniprot.org/

[![Actions](https://img.shields.io/github/actions/workflow/status/althonos/uniprot.rs/test.yml?branch=master&style=flat-square&maxAge=600)](https://github.com/althonos/uniprot.rs/actions)
[![Codecov](https://img.shields.io/codecov/c/gh/althonos/uniprot.rs/master.svg?style=flat-square&maxAge=600)](https://codecov.io/gh/althonos/uniprot.rs)
[![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat-square&maxAge=2678400)](https://choosealicense.com/licenses/mit/)
[![Source](https://img.shields.io/badge/source-GitHub-303030.svg?maxAge=2678400&style=flat-square)](https://github.com/althonos/uniprot.rs)
[![Crate](https://img.shields.io/crates/v/uniprot.svg?maxAge=600&style=flat-square)](https://crates.io/crates/uniprot)
[![Documentation](https://img.shields.io/badge/docs.rs-latest-4d76ae.svg?maxAge=2678400&style=flat-square)](https://docs.rs/uniprot)
[![Changelog](https://img.shields.io/badge/keep%20a-changelog-8A0707.svg?maxAge=2678400&style=flat-square)](https://github.com/althonos/uniprot.rs/blob/master/CHANGELOG.md)
[![GitHub issues](https://img.shields.io/github/issues/althonos/uniprot.rs.svg?style=flat-square&maxAge=600)](https://github.com/althonos/uniprot.rs/issues)


## 🔌 Usage

The [`uniprot::uniprot::parse`](https://docs.rs/uniprot/latest/uniprot/uniprot/fn.parse.html) function
can be used to obtain an iterator over the entries of a UniprotKB database in
XML format (either SwissProt or TrEMBL). XML files for UniRef and UniParc can
also be parsed, with [`uniprot::uniref::parse`](https://docs.rs/uniprot/latest/uniprot/uniref/fn.parse.html)
and [`uniprot::uniparc::parse`](https://docs.rs/uniprot/latest/uniprot/uniparc/fn.parse.html), respectively.

```rust
extern crate uniprot;

let f = std::fs::File::open("tests/uniprot.xml")
   .map(std::io::BufReader::new)
   .unwrap();

for r in uniprot::uniprot::parse(f) {
   let entry = r.unwrap();
   // ... process the Uniprot entry ...
}
```

Any [`BufRead`](https://doc.rust-lang.org/stable/std/io/trait.BufRead.html)
implementor can be used as an input, so the database files can be streamed
directly from their [online location](https://www.uniprot.org/downloads) with
the help of an HTTP library such as [`reqwest`](https://docs.rs/reqwest), or
using the [`ftp`](https://docs.rs/ftp) library.

The XML format is the same for the EBI REST API and for the UniProt API, so
this library can also be used to read single entries or larger queries. For
instance, you can search UniProt for a keyword and retrieve all the matching
entries:

```rust
extern crate ureq;
extern crate libflate;
extern crate uniprot;

let query = "bacteriorhodopsin";
let query_url = format!("https://www.uniprot.org/uniprot/?query={}&format=xml&compress=yes", query);

let req = ureq::get(&query_url).set("Accept", "application/xml");
let reader = libflate::gzip::Decoder::new(req.call().unwrap().into_reader()).unwrap();

for r in uniprot::uniprot::parse(std::io::BufReader::new(reader)) {
    let entry = r.unwrap();
    // ... process the Uniprot entry ...
}
```

See the online documentation at [`docs.rs`](https://docs.rs/uniprot) for more
examples, and some details about the different features available.

## 📝 Features

- [`threading`](https://docs.rs/uniprot/#threading) (_**enabled** by default_):
  compiles the multithreaded parser that offers a 90% speed increase when
  processing XML files.
- [`url-links`](https://docs.rs/uniprot/#url-links) (_**disabled** by default_):
  exposes the `links` in [`OnlineInformation`](https://docs.rs/uniprot/latest/uniprot/uniprot/comment/struct.OnlineInformation.html) as an [`url::Url`](https://docs.rs/url/latest/url/struct.Url.html).

## 🔍 See Also

If you're a bioinformatician and a Rustacean, you may be interested in these
other libraries:

- [`pubchem.rs`](https://github.com/althonos/pubchem.rs): Rust data structures
  and API client for the PubChem API.
- [`obofoundry.rs`](https://github.com/althonos/obofoundry.rs): Rust data
  structures for the OBO Foundry.
- [`fastobo`](https://github.com/fastobo/fastobo): Rust parser and abstract
  syntax tree for Open Biomedical Ontologies.
- [`proteinogenic`](https://github.com/althonos/proteinogenic): Chemical
  structure generation for protein sequences as
  [SMILES](https://en.wikipedia.org/wiki/Simplified_molecular-input_line-entry_system) strings.

## 📋 Changelog

This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
and provides a [changelog](https://github.com/althonos/uniprot.rs/blob/master/CHANGELOG.md)
in the [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) format.


## 📜 License

This library is provided under the open-source
[MIT license](https://choosealicense.com/licenses/mit/).

*This project is in no way not affiliated, sponsored, or otherwise endorsed
by the UniProt Consortium. It was developed
by [Martin Larralde](https://github.com/althonos/) during his PhD project
at the [European Molecular Biology Laboratory](https://www.embl.de/) in
the [Zeller team](https://github.com/zellerlab).*


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


## [Unreleased]
[Unreleased]: https://github.com/althonos/uniprot.rs/compare/v0.7.0...HEAD


## [v0.7.0] - 2023-10-18
[v0.7.0]: https://github.com/althonos/uniprot.rs/compare/v0.6.0...v0.7.0

### Added
- `smartstring` feature for using the `smartstring` crate to reduce heap allocations.

### Changed
- Reduce default sleep duration to reduce strain on CPU.
- Update `quick-xml` dependency to `v0.30.0`.
- Use a dedicated producer thread to read data from the reader in `ThreadedParser`.

### Fixed
- Broken extraction of names in `Citation::from_xml`

### Removed
- Unused `fnv` dependency.
- Deprecated `uniprot::parse` top-level function.


## [v0.6.0] - 2022-10-17
[v0.6.0]: https://github.com/althonos/uniprot.rs/compare/v0.5.2...v0.6.0

### Removed
- Deprecated `CalciumBindingRegion`, `MetalIonBindingSite` and `NucleotidePhosphateBindingRegion` variants of `uniprot::FeatureType`.

### Added
- `uniprot::Ligand` and `uniprot::LigandPart` structs for the `ligand` and `ligand_part` attributes of `uniprot::Feature`.

### Changed
- Updated URLs in documentation examples to use the new Uniprot REST API.
- Moved the `uniref::parse_entry` example to` uniref::parse` since single-entry requests to UniRef don't return single entries anymore.


## [v0.5.2] - 2022-02-28
[v0.5.2]: https://github.com/althonos/uniprot.rs/compare/v0.5.1...v0.5.2

### Added
- `PartialEq`, `Eq`, `Hash` and `Clone` traits to simple enum types (like `uniprot::uniprot::FeatureType`).

### Changed
- Feature gate the `url` crate dependency to skip parsing links into `url::Url` if not needed.
- Remove dependency on `thiserror` by manually implementing `std::error::Error` where needed.


## [v0.5.1] - 2022-01-11
[v0.5.1]: https://github.com/althonos/uniprot.rs/compare/v0.5.0...v0.5.1

### Fixed
- Large test files being included in distributed `crates.io` source package.


## [v0.5.0] - 2022-01-11
[v0.5.0]: https://github.com/althonos/uniprot.rs/compare/v0.4.0...v0.5.0

### Added
- `parse_entry` functions to parse a single UniProt, UniParc or UniRef entry.

### Fixed
- Parsing of creation dates with a defined timezone (e.g. `2021-01-11Z`).


## [v0.4.0] - 2021-07-24
[v0.4.0]: https://github.com/althonos/uniprot.rs/compare/v0.3.1...v0.4.0

### Added
- `uniprot::uniref` module to parse UniRef XML files.
- `uniprot::uniparc` module to parse UniParc XML files.

### Changed
- Moved types to parse UniProt XML files to the `uniprot::uniprot` module.

### Fixed
- Parsers now check the name of the root element before starting to parse the entries.


## [v0.3.1] - 2020-01-19
[v0.3.1]: https://github.com/althonos/uniprot.rs/compare/v0.3.0...v0.3.1

### Changed
- `lazy_static` and `num_cpus` are only required ot build with `threading` feature.
- Slightly improved performance of `ThreadedParser`.


## [v0.3.0] - 2020-01-19
[v0.3.0]: https://github.com/althonos/uniprot.rs/compare/v0.2.0...v0.3.0

### Added
- [`ThreadedParser::with_threads`](https://docs.rs/uniprot/latest/uniprot/parser/struct.ThreadedParser.html#methods) constructor to control the number of threads to spawn when parsing
### Changed
- `ThreadedParser` does not required the reader to be `Send + 'static` anymore.


## [v0.2.0] - 2020-01-18
[v0.2.0]: https://github.com/althonos/uniprot.rs/compare/v0.1.1...v0.2.0

### Added
- Implemented multithreading parser using [`crossbeam-channel`](https://docs.rs/crossbeam-channel), which can be removed by disabling the `threading` feature.
- Improved documentation of `::error` and `::parser` modules.
### Fixed
- Missing implementation of `submittedName` deserialization within `protein` entries that crashed on TrEMBL.


## [v0.1.1] - 2020-01-15
[v0.1.1]: https://github.com/althonos/uniprot.rs/compare/v0.1.0...v0.1.1

### Changed
- Removed remaining explicit [`panic!`](https://doc.rust-lang.org/std/macro.panic.html) calls.
### Added
- [`FromStr`](https://doc.rust-lang.org/std/str/trait.FromStr.html) implementation for some enum types that are read from XML attributes.


## [v0.1.0] - 2020-01-15
[v0.1.0]: https://github.com/althonos/uniprot.rs/compare/52e6940...v0.1.0

Initial release.


src/lib.rs
//! [![Star me](https://img.shields.io/github/stars/althonos/uniprot.rs.svg?style=social&label=Star&maxAge=3600)](https://github.com/althonos/uniprot.rs/stargazers)
//!
//! *Rust data structures and parser for the [UniprotKB database(s)].*
//!
//! [UniprotKB database(s)]: https://www.uniprot.org/
//!
//! [![Actions](https://img.shields.io/github/workflow/status/althonos/uniprot.rs/Test?style=flat-square&maxAge=600)](https://github.com/althonos/uniprot.rs/actions)
//! [![Codecov](https://img.shields.io/codecov/c/gh/althonos/uniprot.rs/master.svg?style=flat-square&maxAge=600)](https://codecov.io/gh/althonos/uniprot.rs)
//! [![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat-square&maxAge=2678400)](https://choosealicense.com/licenses/mit/)
//! [![Source](https://img.shields.io/badge/source-GitHub-303030.svg?maxAge=2678400&style=flat-square)](https://github.com/althonos/uniprot.rs)
//! [![Crate](https://img.shields.io/crates/v/uniprot.svg?maxAge=600&style=flat-square)](https://crates.io/crates/uniprot)
//! [![Documentation](https://img.shields.io/badge/docs.rs-latest-4d76ae.svg?maxAge=2678400&style=flat-square)](https://docs.rs/uniprot)
//! [![Changelog](https://img.shields.io/badge/keep%20a-changelog-8A0707.svg?maxAge=2678400&style=flat-square)](https://github.com/althonos/uniprot.rs/blob/master/CHANGELOG.md)
//! [![GitHub issues](https://img.shields.io/github/issues/althonos/uniprot.rs.svg?style=flat-square&maxAge=600)](https://github.com/althonos/uniprot.rs/issues)
//!
//! # 🔌 Usage
//!
//! All `parse` functions takes a [`BufRead`] implementor as the input.
//! Additionaly, if compiling with the [`threading`] feature, it will
//! require the input to be [`Send`] and `'static` as well. They will use
//! the [`uniprot::Parser`], which is either [`SequentialParser`] or
//! [`ThreadedParser`] depending on the compilation features.
//!
//! ## 🗄️ Databases
//!
//! ### UniProt
//!
//! The [`uniprot::uniprot::parse`] function can be used to obtain an iterator
//! over the entries ([`uniprot::uniprot::Entry`]) of a UniprotKB database in
//! XML format (either [SwissProt] or [TrEMBL]).
//!
//! ```rust
//! extern crate uniprot;
//!
//! let f = std::fs::File::open("tests/uniprot.xml")
//!    .map(std::io::BufReader::new)
//!    .unwrap();
//!
//! for r in uniprot::uniprot::parse(f) {
//!    let entry = r.unwrap();
//!    // ... process the UniProt entry ...
//! }
//! ```
//!
//! The XML format is compatible with the results returned by the UniProt API,
//! so you can also use the [`uniprot::uniprot::parse`] to parse search results:
//!
//! ```rust
//! extern crate ureq;
//! extern crate libflate;
//! extern crate uniprot;
//!
//! let query = "colicin";
//! let req = ureq::get("https://rest.uniprot.org/uniprotkb/search")
//!     .set("Accept", "application/xml")
//!     .query("query", &format!("reviewed:true AND {}", query))
//!     .query("format", "xml")
//!     .query("compress", "true");
//! let reader = libflate::gzip::Decoder::new(req.call().unwrap().into_reader()).unwrap();
//!
//! for r in uniprot::uniprot::parse(std::io::BufReader::new(reader)) {
//!     let entry = r.unwrap();
//!     // ... process the Uniprot entry ...
//! }
//! ```
//!
//! ### UniRef
//!
//! The [`uniprot::uniref::parse`] function can be used to obtain an iterator
//! over the entries ([`uniprot::uniref::Entry`]) of a UniRef database in XML
//! format ([UniRef100], [UniRef90], or [UniRef50]).
//!
//! ### UniParc
//!
//! The [`uniprot::uniparc::parse`] function can be used to obtain an iterator
//! over the entries ([`uniprot::uniparc::Entry`]) of a UniParc database in
//! XML format.
//!
//! ## 📦 Decoding Gzip
//!
//! If parsing a Gzipped file, you can use [`flate2::read::GzDecoder`] or
//! [`libflate::gzip::Decoder`] to decode the input stream, and then simply
//! wrap it in a [`BufferedReader`]. Note that [`flate2`] has slightly better
//! performance, but binds to C,, while [`libflate`] is a pure Rust
//! implementation.
//!
//! ## 📧 Downloading from FTP
//!
//! Uniprot is available from the two following locations: [ftp.ebi.ac.uk]
//! and [ftp.uniprot.org], the former being located in Europe while the
//! latter is in the United States. The `ftp` crate can be used to open
//! a connection and parse the databases on-the-fly: see the
//! [`uniprot::uniprot::parse`] example to see a code snippet.
//!
//! ## 📧 Downloading from HTTP
//!
//! If FTP is not available, note that the EBI FTP server can also be reached
//! using HTTP at [http://ftp.ebi.ac.uk]. This allows using HTTP libraries
//! instead of FTP ones to reach the release files.
//!
//!
//! # 📝 Features
//!
//! ## `threading` - _**enabled** by default_.
//!
//! The `threading` feature compiles the parser module in multi-threaded mode.
//! This feature greatly improves parsing speed and efficiency, but removes
//! any guarantee about the order the entries are yielded in.
//!
//! ## 📋 Changelog
//!
//! This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)
//! and provides a [changelog](https://github.com/althonos/uniprot.rs/blob/master/CHANGELOG.md)
//! in the [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) format.
//!
//! ## 📜 License
//!
//! This library is provided under the open-source
//! [MIT license](https://choosealicense.com/licenses/mit/).
//!
//! [http://ftp.ebi.ac.uk]: http://ftp.ebi.ac.uk
//! [ftp.ebi.ac.uk]: ftp://ftp.ebi.ac.uk
//! [ftp.uniprot.org]: ftp://ftp.uniprot.org
//! [`threading`]: #threading
//! [`flate2`]: https://docs.rs/flate2/
//! [`flate2::read::GzDecoder`]: https://docs.rs/flate2/latest/flate2/read/struct.GzDecoder.html
//! [`libflate`]: https://docs.rs/libflate/
//! [`libflate::gzip::Decoder`]: https://docs.rs/libflate/latest/libflate/gzip/struct.Decoder.html
//! [`BufRead`]: https://doc.rust-lang.org/std/io/trait.BufRead.html
//! [`BufferedReader`]: https://doc.rust-lang.org/std/io/struct.BufReader.html
//! [`Entry`]: ./model/struct.Entry.html
//! [`uniprot::uniprot::parse`]: ./uniprot/fn.parse.html
//! [`uniprot::uniref::parse`]: ./uniref/fn.parse.html
//! [`uniprot::uniparc::parse`]: ./uniparc/fn.parse.html
//! [`uniprot::uniprot::Entry`]: ./uniprot/struct.Entry.html
//! [`uniprot::uniref::Entry`]: ./uniref/struct.Entry.html
//! [`uniprot::uniparc::Entry`]: ./uniparc/struct.Entry.html
//! [`uniprot::Parser`]: ./type.Parser.html
//! [`SequentialParser`]: ./parser/struct.SequentialParser.html
//! [`ThreadedParser`]: ./parser/struct.ThreadedParser.html
//! [SwissProt]: https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.xml.gz
//! [TrEMBL]: https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_trembl.xml.gz
//! [UniRef100]: https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref100/uniref100.xml.gz
//! [UniRef90]: https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref90/uniref90.xml.gz
//! [UniRef50]: https://ftp.uniprot.org/pub/databases/uniprot/uniref/uniref50/uniref50.xml.gz

#![allow(unused_imports)]

extern crate chrono;
#[cfg(feature = "threading")]
extern crate crossbeam_channel;
#[macro_use]
#[cfg(feature = "threading")]
extern crate lazy_static;
#[cfg(feature = "threading")]
extern crate num_cpus;
extern crate quick_xml;
#[cfg(feature = "smartstring")]
extern crate smartstring;
#[cfg(feature = "url-links")]
extern crate url;

#[macro_use]
pub mod parser;
mod common;
pub mod error;
pub mod uniparc;
pub mod uniprot;
pub mod uniref;
